### ISSUE TITLE
Support for multiple images inputs？

### ISSUE THREAD

**Issue by 2kxx** [showlab/Show-o#99]
Note that show-o2 is trained on the interleaved dataset. Does it have the capability of handling multiple image inputs for understanding tasks?

**Comment by Sierkinhane** [comment]
Hi, @2kxx, I guessed we've discussed this via email. @Angusliuuu We use almost the same training data as llava-ov. I think it can support multiple images as input for understanding.

**Comment by Sierkinhane** [comment]
@2kxx Great! It would be really helpful if you could share your code with @Angusliuuu or open a PR, or share it with me so that I can integrate it into our codebase. We’ll make sure to acknowledge your contribution in the repo.

**Comment by 2kxx** [comment]
That's roughly it. My computer isn’t with me, so I recreated a version from memory on my tablet. You can try running it and see.
`# coding=utf-8
import os
from PIL import Image
import wandb
import torch
from tqdm import tqdm
from accelerate.logging import get_logger
from models import Showo2Qwen2_5, omni_attn_mask_naive
from models.misc import get_text_tokenizer
from utils import get_config, flatten_omega_conf, denorm, get_hyper_params, path_to_llm_name, load_state_dict
from datasets.utils import image_transform

logger = get_logger(__name__, log_level="INFO")

if __name__ == '__main__':
    config = get_config()

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    weight_type = torch.float32

    # 初始化 VAE
    if config.model.vae_model.type == 'wan21':
        from models import WanVAE
        vae_model = WanVAE(vae_pth=config.model.vae_model.pretrained_model_path,
                           dtype=weight_type, device=device)
    else:
        raise NotImplementedError

    # 初始化 Show-o 模型
    text_tokenizer, showo_token_ids = get_text_tokenizer(
        config.model.showo.llm_model_path,
        add_showo_tokens=True,
        return_showo_token_ids=True,
        llm_name=path_to_llm_name[config.model.showo.llm_model_path]
    )
    config.model.showo.llm_vocab_size = len(text_tokenizer)

    if config.model.showo.load_from_showo:
        model = Showo2Qwen2_5.from_pretrained(config.model.showo.pretrained_model_path,
                                             use_safetensors=False).to(device)
    else:
        model = Showo2Qwen2_5(**config.model.showo).to(device)
        state_dict = load_state_dict(config.model_path)
        model.load_state_dict(state_dict)

    model.to(weight_type)
    model.eval()

    wandb_config = {k: v for k, v in flatten_omega_conf(config, resolve=True)}
    wandb.init(project="demo", name=config.experiment.name, config=wandb_config)


    fixed_prompt = "What is the difference between the two images?"
    image_path1 = "./mmu_validation/dog.png"
    image_path2 = "./mmu_validation/sofa_under_water.jpg"

    sys_prompt_ids = text_tokenizer("system\nYou are a helpful assistant.<|im_end|>", add_special_tokens=False)['input_ids']
    role_a = text_tokenizer("\n<|im_start|>user\n", add_special_tokens=False)['input_ids']
    role_b = text_tokenizer("\n<|im_start|>assistant\n", add_special_tokens=False)['input_ids']


    image_ori1 = Image.open(image_path1).convert("RGB")
    image1 = image_transform(image_ori1, resolution=config.dataset.preprocessing.resolution).to(device).unsqueeze(0)
    latent1 = vae_model.sample(image1.unsqueeze(2)).squeeze(2).to(weight_type)

    image_emb_und1 = model.image_embedder_und(latent1)
    image_emb_gen1 = model.image_embedder_gen(latent1)
    image_emb_und1 = image_emb_und1 + model.position_embedding(model.image_position_ids)
    image_emb_und1 = model.und_trans(image_emb_und1)['last_hidden_state']
    image_embeds1 = model.fusion_proj(torch.cat([image_emb_und1, image_emb_gen1], dim=-1))

    image_ori2 = Image.open(image_path2).convert("RGB")
    image2 = image_transform(image_ori2, resolution=config.dataset.preprocessing.resolution).to(device).unsqueeze(0)
    latent2 = vae_model.sample(image2.unsqueeze(2)).squeeze(2).to(weight_type)

    image_emb_und2 = model.image_embedder_und(latent2)
    image_emb_gen2 = model.image_embedder_gen(latent2)
    image_emb_und2 = image_emb_und2 + model.position_embedding(model.image_position_ids)
    image_emb_und2 = model.und_trans(image_emb_und2)['last_hidden_state']
    image_embeds2 = model.fusion_proj(torch.cat([image_emb_und2, image_emb_gen2], dim=-1))

    # interleaved
    text_ids1 = text_tokenizer(f"Here is the first image.\n", add_special_tokens=False).input_ids
    text_embed1 = model.showo.model.embed_tokens(torch.tensor(text_ids1).to(device)[None, :])
    text_ids2 = text_tokenizer(f"Here is the second image.\n", add_special_tokens=False).input_ids
    text_embed2 = model.showo.model.embed_tokens(torch.tensor(text_ids2).to(device)[None, :])
    prompt_ids = text_tokenizer(fixed_prompt, add_special_tokens=False).input_ids

    text_tokens_a = torch.tensor([showo_token_ids['bos_id']] + sys_prompt_ids + role_a).to(device)[None, :]
    text_embeds_a = model.showo.model.embed_tokens(text_tokens_a)

    text_tokens_b = torch.tensor([showo_token_ids['boi_id'], showo_token_ids['eoi_id']] + prompt_ids + role_b).to(device)[None, :]
    text_embeds_b = model.showo.model.embed_tokens(text_tokens_b)

    input_embeds = torch.cat([
        text_embeds_a,
        text_embeds_b[:, :1],
        image_embeds1,
        text_embed1,
        image_embeds2,
        text_embed2,
        text_embeds_b[:, 1:]
    ], dim=1).to(weight_type)
    modality_positions = torch.tensor([
        [text_embeds_a.shape[1] + 1,
         num_mmu_image_tokens,
         text_embed1.shape[1],
         num_mmu_image_tokens]
    ], dtype=torch.long).unsqueeze(1).to(device)

    # multi images
    # prompt_ids = text_tokenizer(fixed_prompt, add_special_tokens=False).input_ids
    # 
    # text_tokens_a = torch.tensor([showo_token_ids['bos_id']] + sys_prompt_ids + role_a).to(device)[None, :]
    # text_embeds_a = model.showo.model.embed_tokens(text_tokens_a)
    # 
    # text_tokens_b = torch.tensor([showo_token_ids['boi_id'], showo_token_ids['eoi_id']] + prompt_ids + role_b).to(
    #     device)[None, :]
    # text_embeds_b = model.showo.model.embed_tokens(text_tokens_b)
    # 
    # input_embeds = torch.cat([
    #     text_embeds_a,
    #     text_embeds_b[:, :1],
    #     image_embeds1,
    #     image_embeds2,
    #     text_embeds_b[:, 1:]
    # ], dim=1).to(weight_type)
    # modality_positions = torch.tensor([
    #     [text_embeds_a.shape[1] + 1, num_mmu_image_tokens + num_mmu_image_tokens]
    # ], dtype=torch.long).unsqueeze(1).to(device)

    attention_mask = omni_attn_mask_naive(
        B=input_embeds.size(0),
        LEN=input_embeds.size(1),
        modalities=modality_positions,
        device=device,
        inverted=True
    ).to(input_embeds.dtype)

    output_tokens = model.mmu_generate(
        input_embeds=input_embeds,
        attention_mask=attention_mask,
        top_k=1,
        max_new_tokens=300,
        eos_token=text_tokenizer.eos_token_id
    )
    output_tokens = torch.stack(output_tokens).squeeze()[None]
    text = text_tokenizer.batch_decode(output_tokens, skip_special_tokens=True)

    all_responses.append(f'Image: {os.path.basename(image_path)}\nAnswer: {text[0]}')
    all_images.append(image)

    images_tensor = torch.cat(all_images, dim=0)
    images_tensor = denorm(images_tensor)
    pil_images = [Image.fromarray(img) for img in images_tensor]

    wandb_images = [wandb.Image(img, caption=cap) for img, cap in zip(pil_images, all_responses)]
    wandb.log({"Multimodal responses": wandb_images})
`

**Comment by Angusliuuu** [comment]
ok nb兄弟

**Comment by 2kxx** [comment]
没事没事，两个版本能跑吗，能的话我就发作者一份了
